fasm/parse_fasm.py

```diff
@@ -0,0 +1,217 @@
+
+    return value, tag_to_value_format[tag], i + length
+
+
+def fasm_width_from_bytes(data, i):
+    """ Decode the width of a Verilog value. """
+    if data[i] != tags.width:
+        return None, i
+    else:
+        return int.from_bytes(data[i + 1:i + 5], byteorder), i + 5
+
+
+def fasm_set_feature_from_bytes(data, i):
+    """ Decode a set feature: feature = value """
+    length, i = get_header(tags.set_feature, data, i)
+    if not length:
+        return None, i
+
+    feature, p = tagged_string_from_bytes(tags.feature, data, i)
+    start, end, p = fasm_address_from_bytes(data, p)
+    width, p = fasm_width_from_bytes(data, p)
+    value, value_format, p = fasm_value_from_bytes(data, p)
+
+    assert p == i + length
+
+    assert feature is not None
+
+    if value is None:
+        value = 1
+
+    if width:
+        assert value < 2 ** width, \
+            "value {} larger than specified width of {}".format(value, width)
+
+    return SetFasmFeature(
+        feature=feature,
+        start=start,
+        end=end,
+        value=value,
+        value_format=value_format), i + length
+
+
+def fasm_annotation_from_bytes(data, i):
+    """ Decode an annotation: x = "y" """
+    length, i = get_header(tags.annotation, data, i)
+    if not length:
+        return None, i
+
+    name, p = tagged_string_from_bytes(tags.annotation_name, data, i)
+    value, p = tagged_string_from_bytes(tags.annotation_value, data, p)
+
+    assert p == i + length
+
+    assert name is not None
+    assert value is not None
+
+    return Annotation(name=name, value=value), i + length
+
+
+def fasm_annotations_from_bytes(data, i):
+    """ Decode a set of annotations: { ... } """
+    length, i = get_header(tags.annotations, data, i)
+    if not length:
+        return None, i
+
+    annotations = []
+    annotation, p = fasm_annotation_from_bytes(data, i)
+    while annotation:
+        annotations.append(annotation)
+        annotation, p = fasm_annotation_from_bytes(data, p)
+
+    assert p == i + length
+
+    return annotations, i + length
+
+
+def fasm_line_from_bytes(data, i):
+    """ Decode an entire FASM line. """
+    length, i = get_header(tags.line, data, i)
+    if not length:
+        return None, i
+
+    set_feature, p = fasm_set_feature_from_bytes(data, i)
+    annotations, p = fasm_annotations_from_bytes(data, p)
+    comment, p = tagged_string_from_bytes(tags.comment, data, p)
+
+    assert p == i + length
+
+    assert (
+        set_feature is not None or annotations is not None
+        or comment is not None)
+
+    return FasmLine(
+        set_feature=set_feature, annotations=annotations,
+        comment=comment), i + length
+
+
+def parse_fasm_data(data):
+    """ Parse FASM string, returning list of FasmLine named tuples."""
+    lines = []
+    line, p = fasm_line_from_bytes(data, 0)
+    while line:
+        lines.append(line)
+        line, p = fasm_line_from_bytes(data, p)
+    return lines
```

Comment:
Missing assertion that p == len(data)?

Reply:
`data` does not have a `len()`:

```
TypeError: object of type 'LP_c_char' has no len()
```

Reply:
Just pass the length argument in?  The current code can do a buffer overrun on a mismatch, and that seems like a bad idea.  I've look to see if you can easily add bounds checking to LP_c_char, doesn't look like it :(

Reply:
The current code has a length at the line level, and there is only one line tag. Adding a length argument is equivalent to adding a top level header, which was left out intentionally to allow streaming line by line.

Reply:
Sorry I wasn't clear.  If the python code gets off the rails, and overruns the underlying array, it can segfault/etc.  That is a bad idea?

Reply:
I've added this check.